File: app/src/androidTest/java/com/xmartlabs/template/ui/common/SingleActivityWithRecyclerViewInstrumentationTest.java
package com.xmartlabs.template.ui.common;

import android.app.Activity;
import android.support.annotation.NonNull;
import android.support.test.runner.AndroidJUnit4;
import android.support.v7.widget.RecyclerView;

import com.annimon.stream.Stream;

import org.junit.Test;
import org.junit.runner.RunWith;

import io.appflate.restmock.RESTMockServer;
import io.appflate.restmock.RequestsVerifier;
import io.appflate.restmock.utils.RequestMatcher;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.matcher.ViewMatchers.isAssignableFrom;
import static android.support.test.espresso.matcher.ViewMatchers.isDescendantOfA;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static com.xmartlabs.template.ui.common.RecyclerViewAssertions.countIs;
import static io.appflate.restmock.utils.RequestMatchers.pathContains;
import static org.hamcrest.Matchers.allOf;
import static org.hamcrest.Matchers.anyOf;

/**
 * Created by medina on 21/09/2016.
 */
@RunWith(AndroidJUnit4.class)
public abstract class SingleActivityWithRecyclerViewInstrumentationTest<T extends Activity>
    extends SingleActivityInstrumentationTest<T> {
  protected void checkListSizeTest(@NonNull RecyclerViewItemsTestData testData) {
    RequestMatcher requestMatcher = null;
    if (testData.getServiceUrl() != null) {
      requestMatcher = pathContains("/" + testData.getServiceUrl());
      RESTMockServer
          .whenGET(requestMatcher)
          .thenReturnFile(testData.getJsonResponsePath());
    }

    getActivityTestRule().launchActivity(testData.getIntent());

    Stream.of(testData.getContent())
        .forEach(item ->
            onView(allOf(isAssignableFrom(RecyclerView.class),
                anyOf(withId(item.first), isDescendantOfA(withId(item.first)))))
                .check(matches(countIs(item.second)))
        );

    if (requestMatcher != null) {
      RequestsVerifier.verifyGET(requestMatcher).invoked();
    }
  }

  @Test
  public void hasNoItemsRecyclerView() throws Throwable {
    checkListSizeTest(getNoItemsRecyclerViewItemsTestData());
  }

  @Test
  public void hasItemCountRecyclerView() {
    checkListSizeTest(getItemsRecyclerViewItemsTestData());
  }

  public abstract RecyclerViewItemsTestData getNoItemsRecyclerViewItemsTestData();

  public abstract RecyclerViewItemsTestData getItemsRecyclerViewItemsTestData();
}

File: app/src/androidTest/java/com/xmartlabs/template/ui/common/SingleActivityWithRecyclerViewInstrumentationTest.kt
package com.xmartlabs.template.ui.common

import android.app.Activity
import android.support.test.espresso.Espresso.onView
import android.support.test.espresso.assertion.ViewAssertions.matches
import android.support.test.espresso.matcher.ViewMatchers.*
import android.support.test.runner.AndroidJUnit4
import android.support.v7.widget.RecyclerView
import com.xmartlabs.template.ui.common.RecyclerViewAssertions.countIs
import io.appflate.restmock.RESTMockServer
import io.appflate.restmock.RequestsVerifier
import io.appflate.restmock.utils.RequestMatcher
import io.appflate.restmock.utils.RequestMatchers.pathContains
import org.hamcrest.Matchers.allOf
import org.hamcrest.Matchers.anyOf
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
abstract class SingleActivityWithRecyclerViewInstrumentationTest<T : Activity> : SingleActivityInstrumentationTest<T>() {
  protected fun checkListSizeTest(testData: RecyclerViewItemsTestData) {
    var requestMatcher: RequestMatcher = pathContains("/" + testData.serviceUrl)
    RESTMockServer
        .whenGET(requestMatcher)
        .thenReturnFile(testData.jsonResponsePath)
  
    activityTestRule.launchActivity(testData.intent)
  
    testData.content
        .forEach { item -> onView(allOf(isAssignableFrom(RecyclerView::class.java),
            anyOf(withId(item.first), isDescendantOfA(withId(item.first)))))
            .check(matches(countIs(item.second))) }
  
    RequestsVerifier.verifyGET(requestMatcher).invoked()
  }

  @Test
  @Throws(Throwable::class)
  fun hasNoItemsRecyclerView() {
    checkListSizeTest(noItemsRecyclerViewItemsTestData)
  }

  @Test
  fun hasItemCountRecyclerView() {
    checkListSizeTest(itemsRecyclerViewItemsTestData)
  }

  abstract val noItemsRecyclerViewItemsTestData: RecyclerViewItemsTestData

  abstract val itemsRecyclerViewItemsTestData: RecyclerViewItemsTestData
}
